### PR TITLE
[None][feat] Add CFT counted-write path to MoE all-to-all

### PR DESCRIPTION
**Pull request by zhangcl**
## Summary

Adds counted-write to the MoE all-to-all collective, built on the CUDA Compute Fabric Transport (CFT) API.

The dispatch and combine kernels write directly into peer receive buffers with counted-writes instead of the fence-and-flag handshake. With counted-writes, a counter on the destination will be updated as data arrives.

- `moeAlltoAllCftManager.h` / `moeAlltoAllCftSupport.h` — LE lifecycle (reserve, create, bind, export/import, destroy) and the capability macro
- `moeAlltoAllKernels.{cu,h}` — counted-write dispatch and combine kernels, TMA staging, counter strides
- `thop/moeAlltoAllOp.cpp`, `moeAlltoAllMeta.h` — workspace layout and op plumbing
- `_mnnvl_utils.py` — `CftMnnvlMemory` allocates with `CU_MEM_HANDLE_TYPE_FABRIC` so logical endpoints can bind it; `get_allocation_prop`/granularity become classmethods so the subclass override actually applies
- `moe_alltoall.py`, `nvlink_one_sided.py` — capability detection and per-payload CFT eligibility

## Risk and gating

CFT is **off by default and opt-in**, behind two independent guards:

1. **Runtime.** `can_use_cft_counted_writes` defaults to `False` on both `MoeAlltoAll` and `NVLinkOneSided`. Every existing caller keeps running the fence-based path on every architecture. A caller must opt in explicitly, or set `TRTLLM_MOE_A2A_FORCE_CFT=1`.
2. **Compile time.** `TLLM_CFT_HAS_CUDA_13_4_SUPPORT` requires CUDA 13.4+ and defaults to `0`, so on older toolkits none of the CFT code is built at all.

## Test coverage

Multi-GPU dispatch and combine on GB200 (sm_100, cc 10.0) with `TRTLLM_MOE_A2A_FORCE_CFT=0`, confirming the default fence path is unaffected:

- `tests/unittest/_torch/multi_gpu/test_moe_a2a.py` — **26 passed, 2 skipped**, covering EP4 and EP2, uniform and non-uniform token distributions, `top_k` 2/4/10, the single-token edge case, `use_low_precision_combine` both ways, and the EPLB stats path. The 2 skips are the EP8 parametrizations, which need 8 GPUs on one node.
- `tests/unittest/_torch/distributed/test_moe_a2a_cft.py` — CFT gating logic (`get_force_cft`, `should_use_cft`).

Compile-verified for `sm_100a`, `sm_103a`, and with CFT compiled out (`TLLM_CFT_HAS_CUDA_13_4_SUPPORT=0`).

## Known gaps

- **No multi-GPU test exercises the CFT path itself.** `test_moe_a2a_cft.py` covers the gating predicates only; it is a CPU unit test. CFT-enabled multi-GPU coverage is not included here.
- **EP8 is not covered** by the runtime validation above, as the hardware used has 4 GPUs per node.
- Requesting CFT on a platform whose driver lacks the logical endpoint API currently raises from `moe_a2a_cft_initialize` rather than falling back to the fence path. A capability probe that degrades gracefully is planned as a follow-up.

## Dev Engineer Review

- Adds opt-in CFT counted-write dispatch and combine support for MoE all-to-all.
- Adds logical endpoint lifecycle management, capability checks, handle exchange, and ownership-aware cleanup.
- Adds SM100 counted-write kernels with TMA staging and `fabric.*` operations.
- Extends workspace metadata, counter management, payload planning, and endpoint configuration.
- Adds FABRIC-handle MNNVL allocation through `CftMnnvlMemory`.
- Fixes MNNVL allocation-granularity caching so POSIX and FABRIC handle classes maintain separate cached values.
- Preserves the fence-based path by default and retains timeout handling.
- Requires CUDA 13.4 or later to compile CFT code.
- Review endpoint ownership, synchronization, counter lifetime, workspace bounds, alignment checks, and CUDA error handling.
- Confirm fallback behavior for unsupported logical endpoint drivers.
- Confirm CFT-disabled builds compile without CUDA 13.4 APIs.
- No configuration or test-list files changed.

## QA Engineer Review

- Added `tests/unittest/_torch/distributed/test_moe_a2a_cft.py`.
- Added tests for `get_force_cft` and `should_use_cft`.
- Tests cover environment parsing, forced and unavailable CFT, batch thresholds, payload eligibility, and consistent selection across standalone and NVLink one-sided implementations.
- No `tests/integration/test_lists/`, `test-db/`, or `qa/` entry covers these tests.
- CFT-enabled multi-GPU execution is not tested.
- EP8 runtime coverage is unavailable.
- Verdict: needs follow-up.

### CODE REVIEW

File: cpp/tensorrt_llm/kernels/communicationKernels/moeAlltoAllCftManager.h [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+                cuMemUnmap(localLe_.backingPtr, localLe_.allocSize);
+                cuMemAddressFree(localLe_.backingPtr, localLe_.allocSize);
+            }
+            if (localLe_.memHandle)
+                cuMemRelease(localLe_.memHandle);
+        }
+
+        localLe_.valid = false;
+        initialized_ = false;
+    }
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply cpp/tensorrt_llm/kernels/communicationKernels/moeAlltoAllCftManager.h:466]: +1

File: cpp/tensorrt_llm/thop/moeAlltoAllOp.cpp [hunk 2]
```diff
@@ -194,6 +229,88 @@ torch::Tensor moeA2AInitializeOp(torch::Tensor const& workspace, int64_t epRank,
...
+
+    TORCH_CHECK(g_cft_manager->exchangeEndpoints(allgatherFn), "CftLeManager: Failed to exchange LE endpoints");
+
+    cudaError_t initErr = cudaDeviceSynchronize();
+    if (initErr != cudaSuccess)
+    {
+        fprintf(stderr, "CftLeManager[rank%d]: cudaDeviceSynchronize after init FAILED: %s\n", (int) epRank,
+            cudaGetErrorString(initErr));
+    }
+    tensorrt_llm::mpi::MpiComm::world().barrier();
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply cpp/tensorrt_llm/thop/moeAlltoAllOp.cpp:311]: +1

File: tensorrt_llm/_torch/modules/fused_moe/communication/nvlink_one_sided.py [hunk 3]
```diff
@@ -348,6 +479,23 @@ def __init__(
...
+                self.workspace,
+                self.mnnvl_mem.local_mem_handle,
+                int(self.workspace.size(1)),
+                self.ep_rank,
+                self.ep_size,
+            )
+            NVLinkOneSided._cft_initialized = True
+            from tensorrt_llm._torch.distributed.moe_alltoall import MoeAlltoAll
+
+            MoeAlltoAll._cft_initialized = True
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply tensorrt_llm/_torch/modules/fused_moe/communication/nvlink_one_sided.py]: +1

File: tests/unittest/_torch/distributed/test_moe_a2a_cft.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    can_use_cft: bool,
+    force_cft: bool | None,
+    runtime_max_tokens_per_rank: int,
+    expected: bool,
+):
+    assert should_use_cft(can_use_cft, force_cft, 128, runtime_max_tokens_per_rank) is expected
+    assert (
+        should_use_cft_standalone(can_use_cft, force_cft, 128, runtime_max_tokens_per_rank)
+        is expected
+    )
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply tests/unittest/_torch/distributed/test_moe_a2a_cft.py:70]: +1

File: cpp/tensorrt_llm/thop/moeAlltoAllOp.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhangcl) [reply cpp/tensorrt_llm/thop/moeAlltoAllOp.cpp:311]: Added the following code to check if epSize=worldSize. Currently we only support where epSize=worldSize. And CFT is only supported by NVLinkOneSided.

```
    auto const& cftComm = tensorrt_llm::mpi::MpiComm::world();
    TORCH_CHECK(static_cast<int64_t>(cftComm.getSize()) == epSize,
        "CFT endpoint exchange requires the communicator size (", cftComm.getSize(), ") to equal epSize (", epSize,
        "). MoE all-to-all with CFT counted writes must run as pure EP.");
```

**Comment by xxi-nv** [comment]
Invite @bobboli to help review the communication kernels.

**Review by YihuiLu512 — APPROVED** [review]
LGTM

**Review by xxi-nv — APPROVED** [review]

File: tests/unittest/_torch/distributed/test_moe_a2a_cft.py
Comment (xxi-nv) [inline tests/unittest/_torch/distributed/test_moe_a2a_cft.py:1]: Could you help to move this test into tests/unittest/_torch/modules/moe?
@YihuiLu512 is helping to reorganize the path; MOE-related tests should be put under the path moe.

Comment (zhangcl) [reply tests/unittest/_torch/distributed/test_moe_a2a_cft.py:1]: will do, thanks!

Comment (zhangcl) [reply tests/unittest/_torch/distributed/test_moe_a2a_cft.py:1]: It's been moved.

**Review by BowenFu — APPROVED** [review]
